### PR TITLE
fix: Remove Duplicate Code

### PR DESCRIPTION
**Pull request by mayank-pathakk**
This PR aims to remove the duplicate codes are mentioned in the issue #1261

### CODE REVIEW

File: src/sema/expression/assign.rs [hunk 1]
```diff
@@ -13,6 +13,22 @@ use solang_parser::diagnostics::Diagnostic;
...
+                diagnostics.push(Diagnostic::error(
+                    *loc,
+                    "cannot assign to immutable outside of constructor".to_string(),
+                ));
+                return Err(());
+            }
+        }
+    }
+    Ok(())
+}
```
Comment (seanyoung) [inline src/sema/expression/assign.rs:30]: Actually adding this diagnostic should not be fatal and cause resolving to fail. Please make this function return nothing.

File: src/sema/expression/assign.rs [hunk 2]
```diff
@@ -13,6 +13,22 @@ use solang_parser::diagnostics::Diagnostic;
 use solang_parser::pt;
 use solang_parser::pt::CodeLocation;
 
+///Helper functions
```
Comment (seanyoung) [inline src/sema/expression/assign.rs:16]: That's not a useful comment. How about `// ensure immutable are only mutated in the constructor`

File: src/sema/expression/assign.rs [hunk 3]
```diff
@@ -123,17 +139,7 @@ pub(super) fn assign_single(
...
-                        if !ns.functions[function_no].is_constructor() {
-                            diagnostics.push(Diagnostic::error(
-                                *loc,
-                                "cannot assign to immutable outside of constructor".to_string(),
-                            ));
-                            return Err(());
-                        }
-                    }
-                }
+                check_immutable(&context, ns, loc, &immutable, diagnostics)?;
```
Comment (seanyoung) [inline src/sema/expression/assign.rs:142]: .. and then remove the `?` from the call

**Review by seanyoung — CHANGES_REQUESTED** [review]
Looks good, please make this change while refactoring the code

Thanks!

File: src/sema/expression/assign.rs (hunk 1 above)
Comment (mayank-pathakk) [reply src/sema/expression/assign.rs:28]: Hi @seanyoung ! Thanks for the review :)
I have fixed the code, according to your suggestions.
If the rest of it is okay, I shall start working on the other refactorings.

**Comment by xermicus** [comment]
This PR is front-running  #1274. However it looks like  #1274 got stale

**Comment by xermicus** [comment]
The author of #1274 actually wants to complete his work. First comes first serves.

**Pull request closed without merging**